AsyncDisplayKit/ASRunLoopQueue.mm

```diff
@@ -0,0 +1,98 @@
+//
+//  ASRunLoopQueue.m
+//  AsyncDisplayKit
+//
+//  Created by Rahul Malik on 3/7/16.
+//  Copyright © 2016 Facebook. All rights reserved.
+//
+
+#import "ASRunLoopQueue.h"
+#import "ASThread.h"
+
+#import <deque>
+
+@interface ASRunLoopQueue () {
+  CFRunLoopRef _runLoop;
+  CFRunLoopObserverRef _runLoopObserver;
+  std::deque<id> __renderQueue;
+  ASDN::RecursiveMutex __renderQueueLock;
+}
+@property (nonatomic, copy) void (^queueConsumer)(id dequeuedItem, BOOL isQueueDrained);
+
+@end
+
+@implementation ASRunLoopQueue
+
+- (instancetype)initWithRunLoop:(CFRunLoopRef)runloop andHandler:(void(^)(id dequeuedItem, BOOL isQueueDrained))handlerBlock
+{
+  if (self = [super init]) {
+    _runLoop = runloop;
+    __renderQueue = std::deque<id>();
+    _queueConsumer = [handlerBlock copy];
+    void (^handlerBlock) (CFRunLoopObserverRef observer, CFRunLoopActivity activity) = ^(CFRunLoopObserverRef observer, CFRunLoopActivity activity) {
+      [self processQueue];
```

Comment:
lol nice.  I like that having this an object allows moving the business logic outside of this block.